jacodb-analysis/src/main/kotlin/org/jacodb/analysis/engine/IfdsUnitRunnerFactory.kt

```diff
@@ -0,0 +1,173 @@
+                    is EdgeForOtherRunnerQuery -> {
+                        if (unitResolver.resolve(event.edge.method) == unit) {
+                            backwardRunner?.submitNewEdge(event.edge)
+                        } else {
+                            manager.handleEvent(event, this@BidiIfdsUnitRunner)
+                        }
+                    }
+                    is NewSummaryFact -> {
+                        manager.handleEvent(event, this@BidiIfdsUnitRunner)
+                    }
+                    is QueueEmptinessChanged -> {
+                        forwardQueueIsEmpty = event.isEmpty
+                        val newEvent = QueueEmptinessChanged(backwardQueueIsEmpty && forwardQueueIsEmpty)
+                        manager.handleEvent(newEvent, this@BidiIfdsUnitRunner)
+                    }
+                    is SubscriptionForSummaries -> {
+                        manager.handleEvent(event, this@BidiIfdsUnitRunner)
+                    }
+                }
+            }
+        }
+
+        private val backwardManager: IfdsUnitManager<UnitType> = object : IfdsUnitManager<UnitType> {
+            override suspend fun handleEvent(event: IfdsUnitRunnerEvent, runner: IfdsUnitRunner<UnitType>) {
+                when (event) {
+                    is EdgeForOtherRunnerQuery -> {
+                        if (unitResolver.resolve(event.edge.method) == unit) {
+                            forwardRunner?.submitNewEdge(event.edge)
+                        }
+                    }
+                    is NewSummaryFact -> {
+                        manager.handleEvent(event, this@BidiIfdsUnitRunner)
+                    }
+                    is QueueEmptinessChanged -> {
+                        backwardQueueIsEmpty = event.isEmpty
+                        if (!isParallel && event.isEmpty) {
+                            runner.job.cancel()
+                        }
+                        val newEvent =
+                            QueueEmptinessChanged(backwardQueueIsEmpty && forwardQueueIsEmpty)
+                        manager.handleEvent(newEvent, this@BidiIfdsUnitRunner)
+                    }
+
+                    is SubscriptionForSummaries -> {}
+                }
+            }
+        }
+
+        private var backwardRunner: IfdsUnitRunner<UnitType>? = null
+
+        private var forwardRunner: IfdsUnitRunner<UnitType>? = null
+
+        override suspend fun submitNewEdge(edge: IfdsEdge) {
+            forwardRunner?.submitNewEdge(edge)
+        }
+
+        override val job: Job = scope.launch(start = CoroutineStart.LAZY) {
+            backwardRunner = backwardRunnerFactory
+                .newRunner(graph.reversed, backwardManager, unitResolver, unit, startMethods, this)
+
+            forwardRunner = forwardRunnerFactory
+                .newRunner(graph, forwardManager, unitResolver, unit, startMethods, this)
+
+            if (isParallel) {
+                backwardRunner!!.job.start()
```

Comment:
Apart from lazy part it's better to reduce `!!` with local non-nullable val variable